src/java/com/zimbra/oauth/handlers/impl/FacebookContactsImport.java

```diff
@@ -0,0 +1,529 @@
+import com.zimbra.cs.account.DataSource;
+import com.zimbra.cs.account.DataSource.DataImport;
+import com.zimbra.cs.mailbox.Contact;
+import com.zimbra.cs.mailbox.Mailbox;
+import com.zimbra.cs.mime.ParsedContact;
+import com.zimbra.cs.service.mail.CreateContact;
+import com.zimbra.cs.service.util.ItemId;
+import com.zimbra.oauth.handlers.impl.FacebookOAuth2Handler.FacebookConstants;
+import com.zimbra.oauth.models.OAuthInfo;
+import com.zimbra.oauth.utilities.Configuration;
+import com.zimbra.oauth.utilities.LdapConfiguration;
+import com.zimbra.oauth.utilities.OAuth2Constants;
+import com.zimbra.oauth.utilities.OAuthDataSource;
+
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.apache.commons.httpclient.methods.GetMethod;
+
+
+/**
+ * The FacebookContactsImport class.<br>
+ * Used to sync contacts from the Facebook social service.<br>
+ * Source from the original YahooContactsImport class by @author Greg Solovyev.
+ *
+ * @author Zimbra API Team
+ * @package com.zimbra.oauth.handlers.impl
+ * @copyright Copyright © 2018
+ */
+public class FacebookContactsImport implements DataImport {
+
+  /**
+   * The datasource under import.
+   */
+  private final DataSource mDataSource;
+
+
+  /**
+   * Configuration wrapper.
+   */
+  private Configuration config;
+
+  /**
+   * Constructor.
+   *
+   * @param datasource The datasource to set
+   */
+  public FacebookContactsImport(DataSource datasource) {
+    mDataSource = datasource;
+    try {
+      config = LdapConfiguration.buildConfiguration(FacebookConstants.CLIENT_NAME);
+    } catch (final ServiceException e) {
+      ZimbraLog.extensions.info("Error loading configuration for Facebook: %s", e.getMessage());
+      ZimbraLog.extensions.debug(e);
+    }
+  }
+
+
+  @Override
+  public void test() throws ServiceException {
+    // list of contacts to create after parsing Facebook responses
+    final List<ParsedContact> createList = new ArrayList<ParsedContact>();
+    // existing contacts from the datasource folder
+    final Set<String> existingContacts = new HashSet<String>();
+    String respContent = "";
+    try {
+      // fetch contacts
+      final JsonNode jsonResponse = getContactsRequest(buildContactsUrl(null, refresh()));
+      respContent = jsonResponse.toString();
+      if (jsonResponse != null && jsonResponse.isContainerNode()) {
+        // parse contacts if any, and update the createList
+        if (jsonResponse.has("data") && jsonResponse.get("data").isArray()) {
+          final JsonNode jsonContacts = jsonResponse.get("data");
+          parseNewContacts(existingContacts, jsonContacts, "id", createList);
+        } else {
+          ZimbraLog.extensions.debug(
+              "Did not find 'data' element in JSON response object. Response body: %s",
```

Comment:
We also need to check for error response and throw an exception, that way we no Data import failed